blog/_src/posts/2018-08-16-if-you-give-a-mouse-a-macro.md

```diff
@@ -0,0 +1,445 @@
+        a-deep-dark-void)
+      ⇒ Void]]
+  [(_ (f [x (~datum :) τ] ...) e ...+) ≫
+   -----------------------------------
+   [≻ (define f (λ ([x : τ] ...) e ...))]])
+```
+
+Let's break it down.
+
+1. Create a new type, `Void`, to assign definitions.
+2. Create a constant to serve as the canonical value of type `Void`.
+3. Define a new typed form, `define`, given by two cases.
+4. The first case matches `(define x e)`.
+5. For those cases, check the type of the expression `e`, getting its
+   expansion `e-` and type `τ`.
+6. Create a new name, `x-`.
+7. Expand to Racket's `begin`. Unlike `let`, `begin` does not create a new
+   scope; definitions inside a `begin` are also visible in the surrounding
+   context. That behavior is needed for scenarios like this one that expand to
+   multiple definitions.
+8. Turnstile provides `define-typed-variable-rename` which defines a macro that
+   renames `x` to `x-` in the same manner that we saw above. By using a
+   define-like form, the macro has the same scoping rules as `define`, so it
+   will apply to references to `x` in the surround context---exactly what we
+   want.
+9. Define `x-` to refer to the supplied expression. Note that here `define-` is
+   Racket's `define`.
+10. Keep the result of evaluating this form in line with the type by yielding a
+   value of type `Void`.
+11. The second case allows for function definitions by rewriting to a term that
+    matches the pattern for the first case.
+    
+Unfortunately, while this `define` will allow us to write module-level
+definitions, using it in the body of a function causes an error:
+
+```racket
+(define (add2 [x : Int])
+  (define almost (+ x 1))
+  (+ almost 1))
+==> almost: unbound identifier...
+```
+
+Pointing to the reference on the third line. The problem is that our `define`
+and `begin` forms are not interacting in the way we might have hoped.
+
+When we expand the body of the function above, we associate `x` with type `Int`
+then start checking the body, wrapped in a `begin`:
+
+```racket
+(begin
+  (define almost (+ x 1))
+  (+ almost 1))
+```
+
+Consulting the definition of `begin`, we see that it checks/expands each
+sub-expression in seqence. First in the sequence is a use of `define`, yielding
+
+```
+(begin-
+  (define-syntax almost ...)
+  (define- almost- ...)
+  a-deep-dark-void)
+```
+
+Next, it turns to checking/expanding the subsequent `(+ almost 1)`, where it
+will encounter the identifier `almost`. Even though our `define` formed produced
+a binding for `almost` as part of its output, the expander hasn't actually
+analyzed that binding by the time it reaches the reference in the next
+expression.
+
+The problem is a symptom of analyzing the sequence of forms using an ellipses,
+which corresponds to mapping the typechecking/expanding process over each
+individually. The mapping operation stipulates that checking each item is
+independent of checking the others. But, when we add `define` to the
```

Comment:
At a high level, you can say that you need a "fold" instead of a "map". Ie, instead of binding first, and then checking, you need to interleave binding and checking.